File: Python/local.py
##############################################################################
# LOCAL PARTITIONS MULTI-LABEL CLASSIFICATION                                #
# Copyright (C) 2025                                                         #
#                                                                            #
# This code is free software: you can redistribute it and/or modify it under #
# the terms of the GNU General Public License as published by the Free       #
# Software Foundation, either version 3 of the License, or (at your option)  #
# any later version. This code is distributed in the hope that it will be    #
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of     #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General   #
# Public License for more details.                                           #
#                                                                            #
# 1 - Prof PhD Elaine Cecilia Gatto                                          #
# 2 - Prof PhD Ricardo Cerri                                                 #
# 3 - Prof PhD Mauri Ferrandin                                               #
# 4 - Prof PhD Celine Vens                                                   #
# 5 - PhD Felipe Nakano Kenji                                                #
# 6 - Prof PhD Jesse Read                                                    #
#                                                                            #
# 1 = Federal University of São Carlos - UFSCar - https://www2.ufscar.br     #
# Campus São Carlos | Computer Department - DC - https://site.dc.ufscar.br | #
# Post Graduate Program in Computer Science - PPGCC                          # 
# http://ppgcc.dc.ufscar.br | Bioinformatics and Machine Learning Group      #
# BIOMAL - http://www.biomal.ufscar.br                                       # 
#                                                                            # 
# 1 = Federal University of Lavras - UFLA                                    #
#                                                                            # 
# 2 = State University of São Paulo - USP                                    #
#                                                                            # 
# 3 - Federal University of Santa Catarina Campus Blumenau - UFSC            #
# https://ufsc.br/                                                           #
#                                                                            #
# 4 and 5 - Katholieke Universiteit Leuven Campus Kulak Kortrijk Belgium     #
# Medicine Department - https://kulak.kuleuven.be/                           #
# https://kulak.kuleuven.be/nl/over_kulak/faculteiten/geneeskunde            #
#                                                                            #
# 6 - Ecole Polytechnique | Institut Polytechnique de Paris | 1 rue Honoré   #
# d’Estienne d’Orves - 91120 - Palaiseau - FRANCE                            #
#                                                                            #
##############################################################################


import sys
import platform
import os
import io

import joblib

#FolderRoot = os.path.expanduser('~/LocalPartitions/Python')
#os.chdir(FolderRoot)
#current_directory = os.getcwd()
#sys.path.append('..')

import pickle
import time
import importlib

from joblib import dump
import pandas as pd
import numpy as np

from skmultilearn.problem_transform import BinaryRelevance
from sklearn.ensemble import RandomForestClassifier  

import evaluation as eval
importlib.reload(eval)

import measures as ms
importlib.reload(ms)

if __name__ == '__main__':

    """
    train = pd.read_csv("/tmp/lr-GpositiveGO/Dataset/GpositiveGO/CrossValidation/Tr/GpositiveGO-Split-Tr-1.csv")
    valid = pd.read_csv("/tmp/lr-GpositiveGO/Dataset/GpositiveGO/CrossValidation/Vl/GpositiveGO-Split-Vl-1.csv")
    test = pd.read_csv("/tmp/lr-GpositiveGO/Dataset/GpositiveGO/CrossValidation/Ts/GpositiveGO-Split-Ts-1.csv")
    start = 912
    end = 916
    diretorio = "/tmp/lr-GpositiveGO/Local/Split-1"
    """
    
    train = pd.read_csv(sys.argv[1])
    valid = pd.read_csv(sys.argv[2])
    test = pd.read_csv(sys.argv[3])
    start = int(sys.argv[4])
    end = int(sys.argv[5])
    diretorio = sys.argv[6]    
    
    # juntando treino com validação
    train = pd.concat([train,valid],axis=0).reset_index(drop=True)    

    
    print("\n\n%==============================================%")
    #print("train: ", sys.argv[1])
    #print("valid: ", sys.argv[2])
    #print("test: ", sys.argv[3])
    #print("start: ", sys.argv[4])
    #print("end: ", sys.argv[5])
    print("directory: ", sys.argv[6])
    print("%==============================================%\n\n")
        
    # separa atributos e rótulos
    X_train = train.iloc[:, :start]
    Y_train = train.iloc[:, start:]
    X_test = test.iloc[:, :start]
    Y_test = test.iloc[:, start:]

    labels_y_test = list(Y_test.columns)

    Y_train_dense = np.array(Y_train)
    Y_test_dense = np.array(Y_test)

    # parâmetros do classificador base
    random_state = 1234
    n_estimators = 200
    baseModel = RandomForestClassifier(
        n_estimators=n_estimators, 
        random_state=random_state
    )
    
    classifier = BinaryRelevance(baseModel)
    #classifier = BinaryRelevance(classifier=baseModel, require_dense=[True, True])


    # ======= TREINO =======
    start_train_time = time.time()
    classifier.fit(X_train.values, Y_train.values)
    end_train_time = time.time()
    training = end_train_time - start_train_time
    
    
    # ======= PREDIÇÃO BINÁRIA =======
    start_test_time = time.time()
    binary_predictions = classifier.predict(X_test.values)    
    end_test_time = time.time()
    testing_bin = end_test_time - start_test_time


    # ======= PREDIÇÃO DE PROBABILIDADES =======
    start_test_time = time.time()    
    probas_list = classifier.predict_proba(X_test.values)
    end_test_time = time.time()
    testing_proba = end_test_time - start_test_time

    
    # ======= PEGANDO APENAS A PROBABILIDADE DE PERTENCER =======            
    probas_dense = probas_list.toarray()
    probas_df = pd.DataFrame(probas_dense, columns=Y_test.columns)
    binary_predictions_2 = binary_predictions.toarray()
    binary_df = pd.DataFrame(binary_predictions_2, columns=Y_test.columns)

    
    # ======= SALVANDO OS CSVS =======        
    true = os.path.join(diretorio, "y_true.csv")
    binary = os.path.join(diretorio, "y_pred_bin.csv")
    proba = os.path.join(diretorio, "y_pred_proba.csv")   
    test[labels_y_test].to_csv(true, index=False)    
    probas_df.to_csv(proba, index=False)
    binary_df.to_csv(binary, index=False)
    

    # ======= SAVE TIME =======    
    df_timing = pd.DataFrame([[        
        training,
        testing_bin,
        testing_proba
    ]], columns=["training", "testing_bin", "testing_proba"])
    df_timing.to_csv(os.path.join(diretorio, "runtime-python.csv"), index=False)

    
    # ======= SAVE MODEL SIZE =======
    model_buffer = io.BytesIO()
    pickle.dump(classifier, model_buffer)
    model_size_bytes = model_buffer.tell()
    pd.DataFrame({'size': [model_size_bytes]}).to_csv(
        os.path.join(diretorio, "model-size.csv"), index=False
    )

File: R/local.R
cat("\n\n###############################################################")
  cat("\n# RSCRIPT: START EXECUTE LOCAL PARTITIONS                    #")
  cat("\n###############################################################\n\n")


# clean
rm(list=ls())

##############################################################################
# LOCAL PARTITIONS MULTI-LABEL CLASSIFICATION                                #
# Copyright (C) 2025                                                         #
#                                                                            #
# This code is free software: you can redistribute it and/or modify it under #
# the terms of the GNU General Public License as published by the Free       #
# Software Foundation, either version 3 of the License, or (at your option)  #
# any later version. This code is distributed in the hope that it will be    #
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of     #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General   #
# Public License for more details.                                           #
#                                                                            #
# 1 - Prof PhD Elaine Cecilia Gatto                                          #
# 2 - Prof PhD Ricardo Cerri                                                 #
# 3 - Prof PhD Mauri Ferrandin                                               #
# 4 - Prof PhD Celine Vens                                                   #
# 5 - PhD Felipe Nakano Kenji                                                #
# 6 - Prof PhD Jesse Read                                                    #
#                                                                            #
# 1 = Federal University of São Carlos - UFSCar - https://www2.ufscar.br     #
# Campus São Carlos | Computer Department - DC - https://site.dc.ufscar.br | #
# Post Graduate Program in Computer Science - PPGCC                          # 
# http://ppgcc.dc.ufscar.br | Bioinformatics and Machine Learning Group      #
# BIOMAL - http://www.biomal.ufscar.br                                       # 
#                                                                            # 
# 1 = Federal University of Lavras - UFLA                                    #
#                                                                            # 
# 2 = State University of São Paulo - USP                                    #
#                                                                            # 
# 3 - Federal University of Santa Catarina Campus Blumenau - UFSC            #
# https://ufsc.br/                                                           #
#                                                                            #
# 4 and 5 - Katholieke Universiteit Leuven Campus Kulak Kortrijk Belgium     #
# Medicine Department - https://kulak.kuleuven.be/                           #
# https://kulak.kuleuven.be/nl/over_kulak/faculteiten/geneeskunde            #
#                                                                            #
# 6 - Ecole Polytechnique | Institut Polytechnique de Paris | 1 rue Honoré   #
# d’Estienne d’Orves - 91120 - Palaiseau - FRANCE                            #
#                                                                            #
##############################################################################

cat("\n################################")
cat("\n# Set Work Space               #")
cat("\n###############################\n\n")
library(here)
library(stringr)
FolderRoot <- here::here()
FolderScripts <- here::here("R")
setwd(FolderRoot)


cat("\n########################################")
cat("\n# R Options Configuration              #")
cat("\n########################################\n\n")
options(java.parameters = "-Xmx64g")  # JAVA
options(show.error.messages = TRUE)   # ERROR MESSAGES
options(scipen=20)                    # number of places after the comma


cat("\n########################################")
cat("\n# Creating parameters list              #")
cat("\n########################################\n\n")
parameters = list()


cat("\n########################################")
cat("\n# Reading Datasets-Original.csv        #")
cat("\n########################################\n\n")
name = paste0(FolderRoot, "/datasets-original.csv")
datasets <- data.frame(read.csv(name))
parameters$Datasets.List = datasets


cat("\n#####################################")
cat("\n# GET ARGUMENTS FROM COMMAND LINE   #")
cat("\n#####################################\n\n")
args <- commandArgs(TRUE)

config_file <- args[1]

# config_file = "/home/cissagatto/LocalPartitions/config-files/lr-emotions.csv"


parameters$Config.File$Name = config_file
if(file.exists(config_file)==FALSE){
  cat("\n################################################################")
  cat("#\n Missing Config File! Verify the following path:              #")
  cat("#\n ", config_file, "                                            #")
  cat("#################################################################\n\n")
  break
} else {
  cat("\n########################################")
  cat("\n# Properly loaded configuration file!  #")
  cat("\n########################################\n\n")
}


cat("\n########################################")
cat("\n# Config File                          #\n")
config = data.frame(read.csv(config_file))
print(config)
cat("\n########################################\n\n")



cat("\n########################################")
cat("\n# Getting Parameters                   #\n")
cat("\n########################################")
FolderScripts = toString(config$Value[1])
FolderScripts = str_remove(FolderScripts, pattern = " ")
parameters$Config.File$FolderScripts = FolderScripts

dataset_path = toString(config$Value[2])
dataset_path = str_remove(dataset_path, pattern = " ")
parameters$Config.File$Dataset.Path = dataset_path

folderResults = toString(config$Value[3]) 
folderResults = str_remove(folderResults, pattern = " ")
parameters$Config.File$Folder.Results = folderResults

implementation = toString(config$Value[4])
implementation = str_remove(implementation, pattern = " ")
parameters$Config.File$Implementation = implementation

dataset_name = toString(config$Value[5])
dataset_name = str_remove(dataset_name, pattern = " ")
parameters$Config.File$Dataset.Name = dataset_name

number_dataset = as.numeric(config$Value[6])
parameters$Config.File$Number.Dataset = number_dataset

number_folds = as.numeric(config$Value[7])
parameters$Config.File$Number.Folds = number_folds

number_cores = as.numeric(config$Value[8])
parameters$Config.File$Number.Cores = number_cores

ds = datasets[number_dataset,]
parameters$Dataset.Info = ds


cat("\n########################################")
cat("\n# Creating temporary processing folder #")
cat("\n########################################\n\n")
if (dir.exists(folderResults) == FALSE) {dir.create(folderResults)}



cat("\n########################################")
cat("\n# Loading R Sources                    #")
cat("\n########################################\n\n")
source(file.path(FolderScripts, "libraries.R"))
source(file.path(FolderScripts, "utils.R"))


cat("\n###############################\n")
cat("\n# Get directories             #")
cat("\n###############################\n\n")
diretorios <- directories(parameters)
parameters$Directories = diretorios



cat("\n####################################################################")
cat("\n# Checking the dataset tar.gz file                                 #")
cat("\n####################################################################\n\n")
str00 = paste(parameters$Config.File$Dataset.Path, 
              "/", parameters$Config.File$Dataset.Name,
              ".tar.gz", sep = "")
str00 = str_remove(str00, pattern = " ")

if(file.exists(str00)==FALSE){
  
  cat("\n######################################################################")
  cat("\n# The tar.gz file for the dataset to be processed does not exist!    #")
  cat("\n# Please pass the path of the tar.gz file in the configuration file! #")
  cat("\n# The path entered was: ", str00, "                                  #")
  cat("\n######################################################################\n\n")
  break
  
} else {
  
  cat("\n####################################################################")
  cat("\n# tar.gz file of the DATASET loaded correctly!                     #")
  cat("\n####################################################################\n\n")
  
  # COPIANDO
  str01 = paste("cp ", str00, " ", parameters$Directories$FolderDataset, sep = "")
  res = system(str01)
  if (res != 0) {
    cat("\nError: ", str01)
    break
  }
  
  # DESCOMPACTANDO
  str02 = paste("tar xzf ",  parameters$Directories$FolderDataset, "/", ds$Name,
                ".tar.gz -C ",  parameters$Directories$FolderDataset, sep = "")
  res = system(str02)
  if (res != 0) {
    cat("\nError: ", str02)
    break
  }
  
  #APAGANDO
  str03 = paste("rm ",  parameters$Directories$FolderDataset, "/", ds$Name,
                ".tar.gz", sep = "")
  res = system(str03)
  if (res != 0) {
    cat("\nError: ", str03)
    break
  }
  
}




##############################################################################
#
##############################################################################
if(implementation=="rf"){
  
  source(file.path(FolderScripts, "run-rf.R"))
  
  cat("\n\n############################################################")
  cat("\n# RSCRIPT Local RANDOM FORESTS START                     #")
  cat("\n###########################################################\n\n")
  timeFinal <- system.time(results <- run.rf(parameters))  
  
  
  cat("\n\n#####################################################")
  cat("\n# RSCRIPT SAVE RUNTIME                              #")
  cat("\n#####################################################\n\n")
  result_set <- t(data.matrix(timeFinal))
  setwd(parameters$Directories$FolderLocal)
  write.csv(result_set, "Final-Runtime.csv", row.names = FALSE)
  
  
  cat("\n\n###################################################")
  cat("\n# RSCRIPT DELETE                                  #")
  cat("\n###################################################\n\n")
  str5 = paste("rm -r ", parameters$Directories$FolderDataset, sep="")
  print(system(str5))
  
  
  cat("\n\n###################################################################")
  cat("\n# LOCAL: COMPRESS RESULTS                                      #")
  cat("\n#####################################################################\n\n")
  # str3 = paste("tar -zcvf ", parameters$Directories$FolderLocal, "/",
  #              parameters$Dataset.Info$Name, "-results-local.tar.gz ",
  #              parameters$Directories$FolderLocal, sep="")
  # print(system(str3))
  
  str_01 = paste("tar -zcvf ", parameters$Directories$FolderLocal, "/",
                 parameters$Dataset.Info$Name, "-results-local.tar.gz -C ",
                 parameters$Directories$FolderLocal, " .", sep="")
  print(system(str_01))
  
  
  cat("\n\n###################################################################")
  cat("\n# ====> LPC: COPY TO HOME                                     #")
  cat("\n#####################################################################\n\n")
  
  str_02 = parameters$Directories$folderReports
  if(dir.exists(str_02)==FALSE){dir.create(str_02)}
  
  str_03 = paste(parameters$Directories$FolderLocal, "/",
                 dataset_name, "-results-local.tar.gz", sep="")
  
  str_04 = paste("cp ", str_03, " ", str_02, sep="")
  print(system(str_04))
  
  
  # cat("\n\n######################################################")
  # cat("\n# RSCRIPT COPY TO GOOGLE DRIVE                       #")
  # cat("\n######################################################\n\n")
  # origem = parameters$Directories$FolderLocal
  # destino = paste("nuvem:Local/RandomForests/", dataset_name, sep="")
  # comando = paste("rclone -P copy ", origem, " ", destino, sep="")
  # cat("\n", comando, "\n") 
  # a = print(system(comando))
  # a = as.numeric(a)
  # if(a != 0) {
  #   stop("Erro RCLONE")
  #   quit("yes")
  # }
  # 
  
  # cat("\n\n######################################################")
  # cat("\n# RSCRIPT COPY TO GOOGLE DRIVE                       #")
  # cat("\n######################################################\n\n")
  # origem = diretorios$folderDataset
  # destino = paste("nuvem:Datasets/", dataset_name, sep="")
  # comando = paste("rclone -P copy ", origem, " ", destino, sep="")
  # cat("\n", comando, "\n") 
  # a = print(system(comando))
  # a = as.numeric(a)
  # if(a != 0) {
  #   stop("Erro RCLONE")
  #   quit("yes")
  # }
  
} else if(implementation=="clus"){
  
  
  setwd(FolderScripts)
  source("run-clus.R")
  
  cat("\n\n############################################################")
  cat("\n# RSCRIPT Local START                                     #")
  cat("\n############################################################\n\n")
  timeFinal <- system.time(results <- run.clus(parameters))  
  
  
  cat("\n\n#####################################################")
  cat("\n# RSCRIPT SAVE RUNTIME                              #")
  cat("\n#####################################################\n\n")
  result_set <- t(data.matrix(timeFinal))
  setwd(diretorios$folderLocal)
  write.csv(result_set, "Final-Runtime.csv")
  
  
  cat("\n\n####################################################")
  cat("\n# RSCRIPT DELETE                                   #")
  cat("\n####################################################\n\n")
  str5 = paste("rm -r ", parameters$Directories$FolderDataset, sep="")
  print(system(str5))
  
   
  # cat("\n\n######################################################")
  # cat("\n# RSCRIPT COPY TO GOOGLE DRIVE                       #")
  # cat("\n######################################################\n\n")
  # origem = diretorios$folderLocal
  # destino = paste("nuvem:Local/Clus/", dataset_name, sep="")
  # comando = paste("rclone -P copy ", origem, " ", destino, sep="")
  # cat("\n", comando, "\n") 
  # a = print(system(comando))
  # a = as.numeric(a)
  # if(a != 0) {
  #   stop("Erro RCLONE")
  #   quit("yes")
  # }
  
  cat("\n\n###################################################################")
  cat("\n# LOCAL: COMPRESS RESULTS                                      #")
  cat("\n#####################################################################\n\n")
  # str3 = paste("tar -zcvf ", parameters$Directories$FolderLocal, "/",
  #              parameters$Dataset.Info$Name, "-results-local.tar.gz ",
  #              parameters$Directories$FolderLocal, sep="")
  # print(system(str3))
  
  str_01 = paste("tar -zcvf ", parameters$Directories$FolderLocal, "/",
                 parameters$Dataset.Info$Name, "-results-local.tar.gz -C ",
                 parameters$Directories$FolderLocal, " .", sep="")
  print(system(str_01))
  
  
  cat("\n\n###################################################################")
  cat("\n# ====> GPC: COPY TO HOME                                     #")
  cat("\n#####################################################################\n\n")
  
  str_02 = "~/Local-Partitions/Reports/"
  if(dir.exists(str0)==FALSE){dir.create(str0)}
  
  str_03 = paste(parameters$Directories$FolderLocal, "/",
                 dataset_name, "-results-local.tar.gz", sep="")
  
  str_04 = paste("cp ", str_03, " ", str_02, sep="")
  print(system(str4))
  
  
  
} else if(implementation=="mulan"){
  #  working in progress
  
} else {
  #  working in progress
  
}


cat("\n\n#######################################################")
cat("\n# CLEAN                                               #")
cat("\n#######################################################\n\n")
cat("\nDelete folder \n")
str5 = paste("rm -r ", folderResults, sep="")
print(system(str5))


cat("\n\n################################################################")
cat("\n# RSCRIPT SUCCESSFULLY FINISHED                                #")
cat("\n################################################################\n\n")


rm(list = ls())
gc()

###############################################################################
# Please, any errors, contact us: elainececiliagatto@gmail.com                #
# Thank you very much!                                                        #
###############################################################################
